frontend/src/components/Graph.jsx

```diff
@@ -65,133 +44,158 @@ const Set = (props) => {
-          idx + 1,
-          valsX - 4,
-          props.height * graphRatio + (1 - graphRatio) * 0.5 * props.height + 12
-        );
-      });
-
-
-      if (props.linearRegression) {
-        let xSum = 0,
-        ySum = 0,
-        xxSum = 0,
-        xySum = 0;
-      let count = props.dataPoints.length;
-      for (let i = 0, len = count; i < count; i++) {
-        xSum += i;
-        ySum += props.dataPoints[i];
-        xxSum += i * i;
-        xySum += i * props.dataPoints[i];
-      }
-      let slope = (count * xySum - xSum * ySum) / (count * xxSum - xSum * xSum);
-      let intercept = ySum / count - (slope * xSum) / count;
-
-      console.log("slope: " + slope)
-      console.log("intercept: " + intercept);
+      if (dataPoints.length > 0) {
+        interval = (graphRatio * props.width) / dataPoints.length;
 
-      const findPoint = (x) => {
-        return slope * x + intercept;
+        console.log(interval);
+        dataPoints.map((point) => {
+          if (point > maxData) {
+            maxData = point;
+          }
+        });
       }
+    });
+
+    props.dataPoints.map((dataPoints, colorIdx) => {
+      if (dataPoints.length > 0) {
+        let vertTicFactor = 10;
+        let color = "black";
+        if(colorIdx === 1) {
+            color = "blue";
+        }
+        if(colorIdx === 2) {
+            color = "red";
+        }
+        if(colorIdx === 3) {
+            color = "purple";
+        }
+        for (let i = 0; i <= vertTicFactor; i++) {
+          let tic = canvas.getContext("2d");
+          tic.beginPath();
+          tic.moveTo(
+            ((1 - graphRatio) / 2) * props.width - 2,
+            ((1 - graphRatio) / 2) * props.height +
+              (1 - i / vertTicFactor) * graphRatio * props.height
+          );
+          tic.lineTo(
+            ((1 - graphRatio) / 2) * props.width + 2,
+            ((1 - graphRatio) / 2) * props.height +
+              (1 - i / vertTicFactor) * graphRatio * props.height
+          );
+          tic.stroke();
+
+          const pointText = canvas.getContext("2d");
+          pointText.font = "11px Arial";
+          pointText.fillStyle = "black";
+          pointText.fillText(
+            Math.floor((i / vertTicFactor) * maxData),
+            ((1 - graphRatio) / 2) * props.width - 35,
+            ((1 - graphRatio) / 2) * props.height +
+              (1 - i / vertTicFactor) * graphRatio * props.height
+          );
+        }
 
+        dataPoints.map((dataPoint, idx) => {
+
+
+          let coords = [
+            idx * interval,
+            (dataPoint / maxData) * graphRatio * props.height,
+          ];
+
+          let valsX = coords[0] + ((1 - graphRatio) / 2) * props.width;
+
+          let point = canvas.getContext("2d");
+          point.beginPath();
+          point.arc(
+            valsX,
+            props.height - (coords[1] + ((1 - graphRatio) / 2) * props.height),
+            2,
+            0,
+            2 * Math.PI
+          );
```

Comment:
maybe add inline comments for each number, e.g.,
2, // balabala
0, // balabala
2 * Math.PI, // balabala

Reply:
Good idea